[ENH] Add Glossary of terms/abbreviations used in the specification (#152)
bids-standard/bids-specification: src/02-common-principles.md

1.  **Event** - a stimulus or subject response recorded during a task. Each
    event has an onset time and duration. Note that not all tasks will have
    recorded events (e.g., “resting state”).

1.  **Run** - an uninterrupted repetition of data acquisition that has the same
    acquisition parameters and task (however events can change from run to run
    due to different subject response or randomized nature of the stimuli). Run
    is a synonym of a data acquisition.

1.  **`<index>`** - a numeric value, possibly prefixed with arbitrary number of
    0s for consistent indentation, e.g., it is `01` in `run-01` following
    `run-<index>` specification.

1.  **`<label>`** - an alphanumeric value, possibly prefixed with arbitrary
    number of 0s for consistent indentation, e.g., it is `rest` in `task-rest` 
    following `task-<label>` specification.

1.  **`suffix`** - a portion of the file name with `key-value_` pairs (thus after
    the final `_`), right before the **File extension**.

1.  **File extension** - a portion of the the file name after the left-most
    period (`.`) preceded by any other alphanumeric (so `.gitignore` does not have a

>     suffix)

Reviewer: do you mean the right-most period? Also, the `.gitignore` example is not clear to me: Why is the suffix of relevance when describing what a file extension is?

Reviewer 2: Left-most is correct. Following the right-most rule for `X.nii.gz` would get you `.gz`, not `.nii`.

By this definition, `.func.gii`, `.surf.gii` and `.dtseries.nii` are extensions, which I think is appropriate, but we should make sure we're on the same page here, as these are part of derivatives.

Author: re `.gitignore` -- just an example where there is a left-most period but the `gitignore` is not an extension because is not preceded by anything.  May be I should replace it with `.bidsignore`, which is also not a part of the spec but more relevant here?

Author: @effigies  you bring up the point either period itself is a part of file extension, i.e. either it is `.nii.gz` or `nii.gz`?!  
- in my wording above it is without ("... after ...") so might need to be fixed.
- Python seems to also retain period in `splitext`:
```shell
$> python -c 'from os.path import splitext; print(splitext("bla.nii.gz"))'
('bla.nii', '.gz')
```
- text in bids-specification ATM seems to list them with period (IMHO makes it easier to parse when period is there)
- pybids seems to ask for them without period ([might be my influence](https://github.com/bids-standard/pybids/pull/405) attn: @tyarkoni :
```shell
(git)hopa:~/proj/bids/pybids[sqlalchemy]git
$> git grep extension.*nii
bids/layout/README.md:>>> files = layout.get(subject='0[12]', run=1, extension='.nii.gz')
bids/layout/README.md:In the above snippet, we retrieve all files with subject id 1 or 2 and run id 1 (notice that any entity defined in the config file can be used a filtering argument), and with a file extension of .nii.gz. The returned result is a list of named tuples, one per file, allowing direct access to the defined entities as attributes.
bids/layout/config/bids.json:        "sub-{subject}[/ses-{session}]/dwi/sub-{subject}[_ses-{session}][_acq-{acquisition}]_{suffix<dwi>}{extension<bval|bvec|json|nii\\.gz|nii>|nii\\.gz}",
bids/layout/index.py:                    extension=['nii', 'nii.gz'], suffix='bold',
bids/layout/index.py:                    extension=['nii', 'nii.gz'], suffix='dwi',
bids/layout/layout.py:                         extension=['nii.gz', 'nii'])
bids/layout/layout.py:        images = self.get(extension=['nii', 'nii.gz'], scope=scope,
bids/layout/tests/test_layout.py:              'extension': 'nii.gz'}
bids/layout/tests/test_layout.py:              'desc': 'bleargh', 'extension': 'nii.gz'}
bids/layout/tests/test_layout.py:              'extension': 'nii.gz'}
bids/layout/tests/test_layout.py:              'desc': 'bleargh', 'extension': 'nii.gz'}
bids/layout/tests/test_layout.py:                            acquisition='fullbrain', extension='nii.gz')[0]
bids/reports/parsing.py:            iff_file = [f for f in layout.get(extension='nii.gz') if fn in f.path][0]
bids/reports/parsing.py:                    echos = layout.get_echoes(subject=subj, extension='nii.gz',
bids/reports/parsing.py:                                                     extension='nii.gz',
bids/reports/report.py:            niftis = self.layout.get(subject=subject, extension='nii.gz',
bids/reports/tests/test_parsing.py:    niftis = testlayout.get(subject=subj, extension='nii.gz')
bids/variables/io.py:    images = layout.get(return_type='object', extension='nii.gz',
doc/layout/index.rst:    >>> f = layout.get(task='nback', run=1, extension='nii.gz')[0].filename
doc/layout/index.rst:    >>> f = layout.get(task='nback', run=1, extension='nii.gz')[0].filename
examples/pybids tutorial.ipynb:    "layout.get(subject='01', extension='nii.gz', suffix='bold', return_type='filename')"
examples/pybids tutorial.ipynb:       "{'subject': '01', 'run': 1, 'suffix': 'T2w', 'extension': 'nii.gz'}"
1 11544.....................................:Fri 10 May 2019 08:53:36 AM EDT:.
(git)hopa:~/proj/bids/pybids[sqlalchemy]git
$> git describe
0.7.1-256-g52c6a7f
```
so what should it be? ;)

Reviewer 3: I think that the example should be "so `sub-01_task-resting_bold.nii.gz` has the suffix `.nii.gz`"

Author: `suffix` should be `bold`. Extension `nii.gz` in current wording.

NB it has been awhile for this issue/PR, I might be loosing grip on what should be what now ;)

    "resting state" a task. In the context of brain scanning, a task is always
    tied to one data acquisition. Therefore, even if during one acquisition the
    subject performed multiple conceptually different behaviors (with different
    sets of instructions) they will be considered one (combined) task.

1.  **Event** - a stimulus or subject response recorded during a task. Each
    event has an onset time and duration. Note that not all tasks will have
    recorded events (e.g., “resting state”).

1.  **Run** - an uninterrupted repetition of data acquisition that has the same
    acquisition parameters and task (however events can change from run to run
    due to different subject response or randomized nature of the stimuli). Run
    is a synonym of a data acquisition.

1.  **`<index>`** - a numeric value, possibly prefixed with arbitrary number of
    0s for consistent indentation, e.g., it is `01` in `run-01` following
    `run-<index>` specification.

1.  **`<label>`** - an alphanumeric value, possibly prefixed with arbitrary
    number of 0s for consistent indentation, e.g., it is `rest` in `task-rest`
    following `task-<label>` specification.

> 1.  **`suffix`** - a portion of the file name with `key-value_` pairs (thus after
>     the final `_`), right before the **File extension**.

Reviewer: ```suggestion
1.  **`suffix`** - a portion of the file name with `key-value_` pairs (thus after
    the final `_`), right before the **File extension**, e.g., it is `eeg` in
    `sub-05_task-matchingpennies_eeg.vhdr`.
```

> 1.  **File extension** - a portion of the the file name after the left-most
>     period (`.`) preceded by any other alphanumeric (so `.gitignore` does not have a
>     suffix)

Reviewer: ```suggestion
1.  **File extension** - a portion of the the file name after the left-most
    period (`.`) preceded by any other alphanumeric. For example, `.gitignore` does
    not have a file extension, but the file extension of `test.nii.gz` is `.nii.gz`.
    Note that the left-most period is included in the file extension.
```

> Python seems to also retain period in splitext 
> makes it easier to parse when period is there

I agree with these two arguments. But if pybids does it differently, we may go with how pybids does it. I am not sure ...

Reviewer 3: In MATLAB the period is part of the file extension as well. 
```
>> [p, f, x] = fileparts('/a/b/c.d')
p =
    '/a/b'
f =
    'c'
x =
    '.d'
```

Author: I am torn on this one, but I think that so far it seems that PyBIDS is unfortunately the only argument against including `.`... I think it would be relatively easy to support "official" way with minor compatibility shim on pybids side. Additional minor factor against adding it is added cumbersomness of defining regex based matching (now it would need to include into the group the `.` but also then have another group for alternative extensions e.g. `(?P<extension>.(ext1|ext2))` in Python).

Overall I think we better include '.' to stay inline with current wording etc of BIDS official specification

Author: crap, committed the other suggestion with the wrong commit message.  I will accept it now here and then force push rewrite with corrected messages.

    "resting state" a task. In the context of brain scanning, a task is always
    tied to one data acquisition. Therefore, even if during one acquisition the
    subject performed multiple conceptually different behaviors (with different
    sets of instructions) they will be considered one (combined) task.

1.  **Event** - a stimulus or subject response recorded during a task. Each
    event has an onset time and duration. Note that not all tasks will have
    recorded events (e.g., “resting state”).

1.  **Run** - an uninterrupted repetition of data acquisition that has the same
    acquisition parameters and task (however events can change from run to run
    due to different subject response or randomized nature of the stimuli). Run
    is a synonym of a data acquisition.

1.  **`<index>`** - a numeric value, possibly prefixed with arbitrary number of
    0s for consistent indentation, e.g., it is `01` in `run-01` following
    `run-<index>` specification.

1.  **`<label>`** - an alphanumeric value, possibly prefixed with arbitrary
    number of 0s for consistent indentation, e.g., it is `rest` in `task-rest`
    following `task-<label>` specification.

> 1.  **`suffix`** - a portion of the file name with `key-value_` pairs (thus after

Reviewer 2: ```suggestion
1.  **`suffix`** - an alphanumeric value, located after the `key-value_` pairs (thus after
```

Author: Well... Not that many other entries don't define/constrain syntax yet ;-) but ok, let's do it for suffix too ;-)

Author: For some reason there is no ui elements for me to accept the suggestion, I guess will need to do it offline and push

## Participant names and other labels

BIDS allows for custom user-defined `<label>`s and `<index>`es e.g.,
for naming of participants, sessions, acquisition schemes, etc. Note
that they MUST consist only of allowed characters as described in
[Definitions](02-common-principles.md#definitions) above. In `<index>`es
we RECOMMEND using zero padding (e.g., `01` instead of `1` if you have more than
nine subjects) to make alphabetical sorting more intuitive. Note that

> zero padding is NOT RECOMMENDED to maintain their uniqueness.

Reviewer: ```suggestion
"abusing" zero padding to maintain uniqueness of `<index>`es
is NOT RECOMMENDED.
```

Author: not sure about this one: although we can understand intention of using "abusing" here, I am not sure if it makes message any clearer, while introducing a loosely defined verb in quotes.

Reviewer: ```suggestion
zero padding SHOULD NOT be used to merely maintain uniqueness
of `<index>`es.
```

Author: I like it!
